source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,368 @@
+  // client MUST send a Host header field with an empty field-value.
+  //
+  if (header_map.getHostValue().empty()) {
+    return {RejectOrRedirectAction::Reject, UhvResponseCodeDetail::get().InvalidHost};
+  }
+
+  //
+  // Verify that the path and Host/:authority header matches based on the method.
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-5.4:
+  //
+  // When a proxy receives a request with an absolute-form of request-target, the
+  // proxy MUST ignore the received Host header field (if any) and instead replace
+  // it with the host information of the request-target. A proxy that forwards
+  // such a request MUST generate a new Host field-value based on the received
+  // request-target rather than forward the received Host field-value.
+  // ...
+  // If the target URI includes an authority component, then a client MUST send a
+  // field-value for Host that is identical to that authority component,
+  // excluding any userinfo subcomponent and its "@" delimiter (Section 2.7.1).
+  //
+  // TODO(meilya) - should this be implemented here in UHV or the H1 codec?
+  //
+  auto is_connect_method = header_map.method() == header_values_.MethodValues.Connect;
+  auto is_options_method = header_map.method() == header_values_.MethodValues.Options;
+  auto path_is_asterisk = path == "*";
+  auto path_is_absolute = path.at(0) == '/';
+
+  //
+  // HTTP/1.1 allows for a path of "*" when for OPTIONS requests, based on RFC
+  // 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-5.3.4:
+  //
+  // The asterisk-form of request-target is only used for a server-wide OPTIONS
+  // request
+  // ...
+  // asterisk-form  = "*"
+  //
+  if (!is_options_method && path_is_asterisk) {
+    return {RejectOrRedirectAction::Reject, UhvResponseCodeDetail::get().InvalidUrl};
+  }
+
+  //
+  // Step 2: Validate Transfer-Encoding and Content-Length headers.
+  // HTTP/1.1 disallows a Transfer-Encoding and Content-Length headers,
+  // https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2:
+  //
+  // A sender MUST NOT send a Content-Length header field in any message that
+  // contains a Transfer-Encoding header field.
+  //
+  // The http1_protocol_options.allow_chunked_length config setting can
+  // override the RFC compliance to allow a Transfer-Encoding of "chunked" with
+  // a Content-Length set. In this exception case, we remove the Content-Length
+  // header.
+  //
+  if (header_map.TransferEncoding()) {
+    //
+    // CONNECT methods must not contain a Transfer-Encoding, per RFC 7231,
+    // https://tools.ietf.org/html/rfc7231#section-4.3.6:
+    //
+    // A payload within a CONNECT request message has no defined semantics; sending
+    // a payload body on a CONNECT request might cause some existing
+    // implementations to reject the request.
+    //
+    bool is_chunked = absl::EqualsIgnoreCase(header_map.getTransferEncodingValue(),
```

Comment:
I think this will fail if Transfer-Encoding is multivalue, i.e. `chunked, gzip`. An approach similar to this https://github.com/envoyproxy/envoy/blob/5163f3e16cd774a63e833c9f3b3a07a6da9f0fc0/source/extensions/filters/http/compressor/compressor_filter.cc#L545 is needed.